### PR TITLE
sessionctx: make builtinFunctionUsage thread safe

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/33027

Problem Summary:

### What is changed and how it works?

make builtinFunctionUsage thread safe

### Check List

Tests 

- [x] manual test

We run the workload which can reproduced the issue with this tidb branch for about hours, the concurrent write error disappeared as expected.

### Release note

```release-note
None
```

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -3440,8 +3443,22 @@ func (s *session) updateTelemetryMetric(es *executor.ExecStmt) {
 	}
 }
 
+// GetBuiltinFunctionUsage returns the replica of counting of builtin function usage
 func (s *session) GetBuiltinFunctionUsage() map[string]uint32 {
-	return s.builtinFunctionUsage
+	s.functionUsageMu.RLock()
+	defer s.functionUsageMu.RUnlock()
+	replica := make(map[string]uint32)
```
Comment (ngaut) [inline session/session.go:3450]: please remove map creation out of the lock scope.

Comment (Yisaer) [reply session/session.go]: updated.

**Review by qw4990 — COMMENTED** [review]
LGTM, but is there any way to add some tests for it? @Yisaer

**Review by mjonss — APPROVED** [review]
LGTM

Can a test be added as well?

**Comment by Yisaer** [comment]
I'm afraid it's hard to add a unit test for this issue. However, we have already manually re-run the workload which previously exposed the concurrent write bug for hours with this tidb branch in our internal intergration test, the result is fine as expected. @qw4990 @mjonss

**Review by qw4990 — APPROVED** [review]

**Pull request merged**